src/sparse_matrix_builder.rs

```diff
@@ -0,0 +1,252 @@
+        Self {
+            occurrence: 1,
+            row_sum: initial_value,
+            index: 0,
+        }
+    }
+}
+
+#[derive(Debug, Default, Clone)]
+struct Edge {
+    value: f32,
+}
+
+#[derive(Debug, Clone)]
+pub struct SparseMatrixDescriptor {
+    /// First column index for which we creates subgraph
+    pub col_a_id: u8,
+
+    /// First column name
+    pub col_a_name: String,
+
+    /// Second column index for which we creates subgraph
+    pub col_b_id: u8,
+
+    /// Second column name
+    pub col_b_name: String,
+}
+
+impl SparseMatrixDescriptor {
+    pub fn new(col_a_id: u8, col_a_name: String, col_b_id: u8, col_b_name: String) -> Self {
+        Self {
+            col_a_id,
+            col_a_name,
+            col_b_id,
+            col_b_name,
+        }
+    }
+
+    pub fn make_buffer(&self) -> SparseMatrixBuffer {
+        SparseMatrixBuffer {
+            descriptor: self.clone(),
+            edge_count: 0,
+            hash_2_row: Default::default(),
+            hashes_2_edge: Default::default(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct SparseMatrixBuffer {
+    pub descriptor: SparseMatrixDescriptor,
+    pub edge_count: u32,
+    hash_2_row: HashMap<u64, Entity, BuildHasherDefault<FxHasher>>,
+    hashes_2_edge: HashMap<(u64, u64), Edge, BuildHasherDefault<FxHasher>>,
+}
+
+impl SparseMatrixBuffer {
+    pub fn handle_pair(&mut self, hashes: &[u64]) {
+        let a = self.descriptor.col_a_id;
+        let b = self.descriptor.col_b_id;
+        self.add_pair_symmetric(
+            hashes[(a + 1) as usize],
+            hashes[(b + 1) as usize],
+            hashes[0],
+        );
+    }
+
+    fn add_pair_symmetric(&mut self, a_hash: u64, b_hash: u64, count: u64) {
+        let value = 1f32 / (count as f32);
+
+        self.update_row(a_hash, value);
+        self.update_row(b_hash, value);
+
+        self.edge_count += 1;
+
+        self.update_edge(a_hash, b_hash, value);
+        self.update_edge(b_hash, a_hash, value);
+    }
+
+    fn update_row(&mut self, hash: u64, val: f32) {
+        match self.hash_2_row.entry(hash) {
+            std::collections::hash_map::Entry::Vacant(entry) => {
+                entry.insert(Entity::new(val));
+            }
+            std::collections::hash_map::Entry::Occupied(mut entry) => {
+                let row = entry.get_mut();
+                row.occurrence += 1;
+                row.row_sum += val;
+            }
+        }
+    }
+
+    fn update_edge(&mut self, a_hash: u64, b_hash: u64, val: f32) {
+        match self.hashes_2_edge.entry((a_hash, b_hash)) {
+            std::collections::hash_map::Entry::Vacant(entry) => {
+                entry.insert(Edge { value: val });
+            }
+            std::collections::hash_map::Entry::Occupied(mut entry) => {
+                entry.get_mut().value += val;
+            }
+        };
+    }
+
+    pub fn finish(desc: &SparseMatrixDescriptor, buffers: Vec<SparseMatrixBuffer>) -> SparseMatrix {
```

Comment:
It's quite long function, maybe we could split it up into "aggregate_entities" & "aggregate_edges"?

Reply:
Done and done